File: cpp/p345-reverse-vowels-of-string.cpp
#include <algorithm>
#include <iostream>

class Solution {
public:
    std::string reverseVowels(std::string s)
    {
        if (s.empty()) {
            return s;
        }
        const std::string vowels = "AEIOUaeiou";
        size_t i1 = 0;
        size_t i2 = s.size() - 1;
        while (i1 < i2) {
            while (!std::binary_search(vowels.begin(), vowels.end(), s[i1])) {
                ++i1;
                if (i1 >= i2) {
                    return s;
                }
            }
            while (!std::binary_search(vowels.begin(), vowels.end(), s[i2])) {
                --i2;
                if (i1 >= i2) {
                    return s;
                }
            }
            std::swap(s[i1], s[i2]);
            ++i1;
            --i2;
        }
        return s;
    }
};

struct TestCase {
    std::string s;
    std::string exp;
};

int main(int /*argc*/, char* /*argv*/ [])
{
    TestCase testCases[] = {
        { "hello", "holle" },
        { "leetcode", "leotcede" },
        { "HELLO", "HOLLE" },
    };

    Solution s;
    for (const auto& tc : testCases) {
        const auto act = s.reverseVowels(tc.s);
        if (tc.exp != act) {
            std::cerr << "FAIL. "
                      << tc.exp << " != " << act << '\n';
        }
    }
    return 0;
}

File: leetcode/py/p345-reverse-vowels-of-string.py
class Solution:
    def reverseVowels(self, s: str) -> str:
        l = list(s)
        if len(l) <= 1:
            return s
        vowels = {'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'}
        x1 = 0
        x2 = len(l) - 1
        while x1 < x2:
            if l[x1] in vowels:
                if l[x2] in vowels:
                    l[x1], l[x2] = l[x2], l[x1]
                    x1 += 1
                    x2 -= 1
                else:
                    x2 -= 1
            else:
                x1 += 1
        return ''.join(l)


if __name__ == '__main__':

    tests = (
        ("hello", "holle"),
        ("leetcode", "leotcede"),
        ("HELLO", "HOLLE"),
    )

    s = Solution()
    for test in tests:
        string = test[0]
        exp = test[1]
        act = s.reverseVowels(string)
        if exp != act:
            print(f'FAIL. {exp} != {act}')
